File: AddTwoNumbers.cs
using System;
/** https://leetcode.com/problems/add-two-numbers/
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode AddTwoNumbers(ListNode l1, ListNode l2) {
        //ListNode result = new ListNode();
        //ListNode result = null;
        //ListNode resultCurrent = result;
        ListNode resultCurrent = null;
        ListNode result = null;

        ListNode l1Current = l1;
        ListNode l2Current = l2;
        int remainderFromPrevious = 0;
        while(l1Current != null || l2Current != null)
        {
            //get the total
            int total =  remainderFromPrevious;
            if(l1Current != null)
            {
                total += l1Current.val;
            }
            
            if(l2Current != null)
            {
                total += l2Current.val;
            }
            
            
            if(result == null) //starting the list
            {
            
                //make a new node
                resultCurrent = new ListNode();

                //break the total into the next value and the current value            
                //store the current value
                resultCurrent.val = (int)total%10;
                
                result = resultCurrent;
                
                //resultCurrent = resultCurrent.next;
            }
            else
            {
                resultCurrent.next = new ListNode();
                resultCurrent = resultCurrent.next;

                //break the total into the next value and the current value            
                //store the current value
                resultCurrent.val = (int)total%10;
                                
                //resultCurrent = resultCurrent.next;
            }
            
            //pass on the remainder            
            remainderFromPrevious = (int)Math.Floor( (double)(total/10) );
            Console.WriteLine("remainderFromPrevious is now " + remainderFromPrevious);
            
            if(l1Current != null)
            {
                l1Current = l1Current.next;
            }
            if(l2Current != null)
            {
                l2Current = l2Current.next;
            }
        }
        
        if(remainderFromPrevious > 0)
        {
            //if there's a leftover remainder, add a digit for that.
            resultCurrent.next = new ListNode();
            resultCurrent = resultCurrent.next;
            resultCurrent.val = remainderFromPrevious;
        }
        
        return result;
    }
}

File: crackingTheCodingInterview/addTwoNumbers.cpp
//https://leetcode.com/problems/add-two-numbers/submissions/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        
        ListNode* result = new ListNode(0);
        ListNode* resultIter = result;
        int carry = 0;
        while(l1 != nullptr || l2 != nullptr)
        {
            int newTotal = 0;
            
            newTotal += (l1 == nullptr ? 0 : l1->val);
            newTotal += (l2 == nullptr ? 0 : l2->val);
            newTotal += carry;

            resultIter->next = new ListNode( newTotal%10 );
            resultIter = resultIter->next;
            carry = ( newTotal/10 );

            l1 = (l1 == nullptr ? l1 : l1->next);
            l2 = (l2 == nullptr ? l2 : l2->next);
        }
        
        if(carry > 0)
        {
            resultIter->next = new ListNode( carry );
        }
        
        return result->next;
    }
};